### PR TITLE
[ENH] Duet Implementation

### LINKED ISSUE #1938: [ENH] Implementation of DUET
**Is your feature request related to a problem? Please describe.**
Implementation of DUET for forecasting

**Describe the solution you'd like**
Integrated with pytorch forecasting

**Describe alternatives you've considered**
na

**Additional context**

Ref: https://arxiv.org/abs/2412.10859

### PR DESCRIPTION
**Pull request by Shvrth**
#### Reference Issues/PRs

Fixes #1938
#### What does this implement/fix? Explain your changes.

This is to integrate the DUET model with pytorch-forecasting as per issue #1938 
Link to paper: https://arxiv.org/abs/2412.10859
Link to official implementation: https://github.com/decisionintelligence/DUET

- _pytorch_forecasting/models/duet_ contains the model code, and a dataset for testing.
- _tests/test_models/test_duet.py_ contains a simple test script for training the model.

#### What should a reviewer concentrate their feedback on?

- I'm particularly looking for feedback on the implementation of DUET as it's still a work-in-progress.
- Any suggestions for improving clarity or adhering to the project's style.
- I've added a test script for the model, but I'm observing that the training loss stagnates after the first epoch. I'm investigating whether this is a general issue (e.g., needing a learning rate scheduler) or a specific flaw in the DUET architecture. Any insights here would be particularly helpful.

#### Did you add any tests for the change?

I have added a test script for verifying the model's training process at _tests/test_models/test_duet.py_

#### Any other comments?

The current implementation does not yet include the time-based features from the original DUET model. This will be addressed in a future commit.
#### PR checklist

- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [x] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

File: pytorch_forecasting/models/duet/duet.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
+from copy import copy
+from logging import config
+from typing import Callable, Optional, Union
+
+from einops import rearrange
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/duet/duet.py]: Can we use `_safe_import` here, this is not a core dep of `ptf`, so it should be imported using `_safe_import`. See [here](https://github.com/sktime/pytorch-forecasting/blob/21a46a21d1a2c9728736438f0ad2827712ae9188/pytorch_forecasting/utils/_dependencies/_safe_import.py#L14C5-L14C17)

File: pytorch_forecasting/models/duet/duet.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        num_experts: int = 4,
+        noisy_gating: bool = True,
+        k: int = 1,
+        CI: bool = False,
+        **kwargs,
+    ):
+        """
+        Initialize DUET model.
+
+        Args:
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/duet/duet.py:126]: Please use [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) style for docstrings. See [here](https://github.com/sktime/pytorch-forecasting/blob/21a46a21d1a2c9728736438f0ad2827712ae9188/pytorch_forecasting/models/xlstm/_xlstm.py#L46-L70) for example.

File: pytorch_forecasting/models/duet/sub_modules/layers/Autoformer_EncDec.py
Comment (phoeenniixx) [inline pytorch_forecasting/models/duet/sub_modules/layers/Autoformer_EncDec.py:1]: Can you add these to [`pytorch-forecasting/layers`](https://github.com/sktime/pytorch-forecasting/tree/main/pytorch_forecasting/layers) module, rather than in the `duet` folder? There you may find some layers that you could reuse here? Like [`revin`](https://github.com/sktime/pytorch-forecasting/blob/main/pytorch_forecasting/layers/_normalization/_revin.py)?

File: pytorch_forecasting/models/duet/sub_modules/utils/masked_attention.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        A = self.dropout(torch.softmax(scale * scores, dim=-1))
+        V = torch.einsum("bhls,bshd->blhd", A, values)
+
+        if self.output_attention:
+            return V.contiguous(), A
+        else:
+            return V.contiguous(), None
+
+
+class AttentionLayer(nn.Module):
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/duet/sub_modules/utils/masked_attention.py:115]: This should be added to `layers/attention`?

File: pytorch_forecasting/models/duet/sub_modules/utils/__init__.py
Comment (phoeenniixx) [inline pytorch_forecasting/models/duet/sub_modules/utils/__init__.py:1]: Maybe you should add the utils to `utils` module and layers that are present here should go to `layers`?

File: tests/test_models/test_duet.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        print(f"\n--- Epoch {trainer.current_epoch} Summary ---")
+        if metrics:
+            for key, value in metrics.items():
+                print(f"{key}: {value.item():.4f}")
+        else:
+            print("No metrics logged for this epoch.")
+        print("--------------------------------------\n")
+
+
+def configure_dataset() -> tuple[TimeSeriesDataSet, TimeSeriesDataSet]:
```
Comment (PranavBhatP) [inline tests/test_models/test_duet.py:27]: I have a change request w.r.t tests for duet. Could you try creating an `_integration` function to run the entire pipeline starting from the dataset creation all the way to the predictions. You can refer to this example - [test_timexer](https://github.com/sktime/pytorch-forecasting/blob/main/tests/test_models/test_timexer.py).

You can then run individual tests on `DUET` by calling this `_integration` on different kinds of data. You can understand how to do this from the linked example.

Comment (Shvrth) [reply tests/test_models/test_duet.py:27]: Yes, thank you for the feedback. I'll update the code to include this.

Comment (fkiraly) [reply tests/test_models/test_duet.py:27]: @PranavBhatP, would the current test suite not already run the `_integration`, with parameters from the pkg class?

File: pytorch_forecasting/models/duet/sub_modules/utils/masked_attention.py (hunk 3 above)
Comment (fkiraly) [reply pytorch_forecasting/models/duet/sub_modules/utils/masked_attention.py:115]: do we not already have an attention layer?

File: tests/test_models/test_duet.py (hunk 4 above)
Comment (PranavBhatP) [reply tests/test_models/test_duet.py:27]: that also works, but since @Shvrth has already included a test file for duet, I suggested the change since we have a common pattern of testing these models in their respective files. Maybe there are also model specific cases to be tested.

File: pytorch_forecasting/models/duet/sub_modules/utils/masked_attention.py (hunk 3 above)
Comment (phoeenniixx) [reply pytorch_forecasting/models/duet/sub_modules/utils/masked_attention.py:115]: I didn't read the code inside, but I am assuming it may be a slightly different implementation, if yes, we should add it otherwise we could use the existing [one](https://github.com/sktime/pytorch-forecasting/blob/e1cc1ced27833741ad9896954785bfdae4f983f6/pytorch_forecasting/layers/_attention/_attention_layer.py#L13). It depends on @Shvrth's requirements...

**Comment by PranavBhatP** [comment]
@Shvrth are you still working on this? I see the tests are failing. If you are facing any issues, please let us know!
